### PR TITLE
community - Allow the use of the json_metadata 'canonical_link' to set canonical links

### PR DESCRIPTION
**Pull request by drov0**
This is a followup of #336. 

Long story short, the current implementation involves a JSON file which does app (from json_metadata) => cannonical link via a list Which works great for dapps like dtube or busy which have a static domain but doesn't work at all for steempress.

If you are unaware about the project, steempress is a wordpress plugin to allow wordpress powered websites to use steem (cross-posting and using steem as a comment section)

So basically this logic is added : add a parameter in the json_metadata "canonical_link" and if it's set, use that url as canonical link, otherwise fallback to the app logic.

### CODE REVIEW

**Comment by reggaemuffin** [comment]
That is a much needed feature for TokenBB too, as we also allow custom domains for different forums.

File: src/app/utils/CanonicalLinker.js [hunk 1]
```diff
@@ -1,30 +1,40 @@
...
-                .join(d.category)
-                .split('{username}')
-                .join(d.author)
-                .split('{permlink}')
-                .join(d.permlink);
+    let canonicalUrl = 'https://steemit.com' + d.link;
+
+    if (d.json_metadata) {
+        if (
+            d.json_metadata.canonical_link &&
```
Comment (roadscape) [inline src/app/utils/CanonicalLinker.js]: Could we use `canonical_url` instead of `canonical_link`? In this context `link` may imply an object consisting of URL, title, and optional attributes.

File: src/app/utils/CanonicalLinker.js [hunk 2]
```diff
@@ -1,30 +1,40 @@
...
+            typeof d.json_metadata.canonical_link === 'string'
+        ) {
+            const urlTester = new RegExp(
+                /^(?:http(s)?:\/\/)?[\w.-]+(?:\.[\w\.-]+)+[\w\-\._~:/?#[\]@!\$&'\(\)\*\+,;=.]+$/gm
```
Comment (roadscape) [inline src/app/utils/CanonicalLinker.js]: The regex on this line seems to make the protocol optional and it's something that needs to be explicit. In some other places we use a simpler regex, e.g. `/^https?:\/\//`.

Comment (drov0) [reply src/app/utils/CanonicalLinker.js]: Good point I'll use that.

File: src/app/utils/CanonicalLinker.js (hunk 1 above)
Comment (drov0) [reply src/app/utils/CanonicalLinker.js]: Ah I wasn't aware of that, I used the general convention regarding canonical links as "canonical_url" sounded a bit odd. But alright I'll switch to that.

**Comment by drov0** [comment]
Link to a post to test the feature http://localhost:8080/test/@petanque/test-canonical-link1

**Comment by roadscape** [comment]
Local PR: #3435
